### PR TITLE
Add optional binary relocatability

### PR DESCRIPTION
**Pull request by traversaro**
# 🎉 New feature

Closes part of https://github.com/gazebosim/gz-sim/issues/626

## Summary

This PR uses the changes introduced in gz-cmake3 in https://github.com/gazebosim/gz-cmake/pull/334 to support the cmake installation directory to be moved after the `make install` prefix, and continue to work without the need to set any special environment variable, as long as the library is compiled as shared. To avoid regressions and problems in Ubuntu Focal due to the use of std::filesystem, this new behaviour is only activated if the `GZ_ENABLE_RELOCATABLE_INSTALL` option is enabled, and its default value is `OFF` .

In particular, this PR defines a `gz::rendering::getResourcePath()` and `gz::rendering::getEngineInstallDir()` that should be used in place of the `GZ_RENDERING_RESOURCE_PATH` and `GZ_RENDERING_ENGINE_INSTALL_DIR` macros to ensure that the library is relocatable.

Furthermore, this PR also deprecates the `GZ_RENDERING_RESOURCE_PATH`  and `GZ_RENDERING_ENGINE_INSTALL_DIR` macros, using the strategy described in https://stackoverflow.com/a/29297970 . That strategy works fine on GCC and Clang, while on MSVC it raise a warning:
~~~
warning C4068: unknown pragma
~~~

However, I think that it does anyhow the job of raising some kind of warning, and then at soon as the developer checks the macro definition the kind of warning is clear.

## Test it

The test should work as usual. The used CMake machinery is tested in https://github.com/gazebosim/gz-cmake/pull/334 .

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by traversaro** [comment]
The PR is failing as it requires the merge and release of https://github.com/gazebosim/gz-cmake/pull/334 . However, it was important to already open it to permit to understand how https://github.com/gazebosim/gz-cmake/pull/334 would be used.

**Comment by traversaro** [comment]
I updated this PR to target gz-rendering7. The related change in gz-cmake3 has been merged in https://github.com/gazebosim/gz-cmake/pull/334, I guess we need a release of gz-cmake3 before going forward here.

File: src/InstallationDirectories.cc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+inline namespace GZ_RENDERING_VERSION_NAMESPACE {
+
+std::string getResourcePath()
+{
+  return gz::common::joinPaths(getInstallPrefix(), GZ_RENDERING_RELATIVE_RESOURCE_PATH);
+}
+
+std::string getEngineInstallDir()
+{
+  return gz::common::joinPaths(getInstallPrefix(), GZ_RENDERING_ENGINE_RELATIVE_INSTALL_DIR);
```
Comment (iche033) [inline src/InstallationDirectories.cc:36]: CI picked up a couple of minor issues here:

```
  /github/workspace/src/InstallationDirectories.cc:31:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/InstallationDirectories.cc:36:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

Comment (traversaro) [reply src/InstallationDirectories.cc:36]: Thanks! I did not notice that as I tought the CI errors were all due to the gz-cmake 3.1.0 missing on Focal.

Comment (traversaro) [reply src/InstallationDirectories.cc]: Fixed in https://github.com/gazebosim/gz-rendering/pull/804/commits/c4129159e3715e10b31b35342d589eb01a1be38d .

**Review by iche033 — APPROVED** [review]
looks good to me.

@mjcarroll do you need to test this with bazel?

**Comment by mjcarroll** [comment]
I already tested the physics one, since the remainder follow suit, i don't have concerns here.

**Comment by PetervdPerk-NXP** [comment]
This PR triggers following CMake error for me.
Which package should provide the `gz_add_get_install_prefix_impl` cmake command?

```
CMake Error at src/CMakeLists.txt:18 (gz_add_get_install_prefix_impl):
  Unknown CMake command "gz_add_get_install_prefix_impl".
```

**Pull request merged**